### ISSUE TITLE
Query: Derived include types not inferred for collections

### ISSUE THREAD

**Issue by AndriySvyryd** [aspnet/EntityFrameworkCore#10441]
See https://github.com/aspnet/EntityFrameworkCore/compare/Repro10440

Using the desired syntax
``` C#
context.Levels
    .Include(c => c.Items)
    .Include(l => l.Items).ThenInclude((Container c) => c.Items)
```
results in
>The type 'Microsoft.EntityFrameworkCore.SqlServerEndToEndTest.Container' cannot be used as type parameter 'TDerived' in the generic type or method 'EntityFrameworkQueryableExtensions.ThenInclude<TEntity, TPreviousProperty, TDerived, TProperty>(IIncludableQueryable<TEntity, TPreviousProperty>, Expression<Func<TDerived, TProperty>>)'. There is no implicit reference conversion from 'Microsoft.EntityFrameworkCore.SqlServerEndToEndTest.Container' to 'System.Collections.Generic.ICollection<Microsoft.EntityFrameworkCore.SqlServerEndToEndTest.Item>'

Have to specify types explicitly 😞 :
``` C#
context.Levels
    .Include(c => c.Items)
    .Include(l => l.Items).ThenInclude<Level, Item, Container, IEnumerable<Item>>(c => c.Items)
```

**Comment by divega** [comment]
Initial notes: 

It seems that the overload of `ThenInclude` that supports navigating to properties of derived elements of collections works fine per se, and it is the presence of the overload for derived references what is interfering with overload resolution. E.g.

If I comment out the reference version of the overload:

``` csharp
    public static IIncludableQueryable<TEntity, TProperty> ThenInclude
        <TEntity, TPreviousProperty, TDerived, TProperty>(
            [NotNull] this IIncludableQueryable<TEntity, TPreviousProperty> source,
            [NotNull] Expression<Func<TDerived, TProperty>> navigationPropertyPath)
            where TEntity : class
            where TDerived : TPreviousProperty
```

Then the test compiles and the collection overload is selected:

``` csharp
    public static IIncludableQueryable<TEntity, TProperty> ThenInclude
        <TEntity, TPreviousProperty, TDerived, TProperty>(
            [NotNull] this IIncludableQueryable<TEntity, IEnumerable<TPreviousProperty>> source,
            [NotNull] Expression<Func<TDerived, TProperty>> navigationPropertyPath)
            where TEntity : class
            where TDerived : TPreviousProperty
```

**Comment by AndriySvyryd** [comment]
Should we rename one of them? `ThenIncludeMany`?

**Comment by smitpatel** [comment]
This looks somewhat similar to #4117 which faces issue with reference/collection overloads. #4117 was limited to the case of intellisense showing matches from only 1 overload. But compiler should still be able to find correct overload and not fail like this issue.

**Comment by divega** [comment]
**Note for triage:**

A few of us discussed the following action items:

1. @divega to create a compiler issue to give the compiler team a chance to figure out if the current behavior is by design (seems a bit strange that he compiler is selecting a non-applicable overload when an applicable one exists, however there is type inference in the mix, which complicates things).

2. @maumar remove the overloads of `Include` and `ThenInclude` with `TDerived` to make sure they are not there for the first preview of 2.1

3. Investigate supporting expressions with casting (strong cast and/or `as` operator). 

Re the latter, here are a couple of examples:

*Using the `as` operator:*
``` csharp
 var level = context.Levels
    .Include(l => l.Items).ThenInclude(c => (c as Container).Items)
    .ToList().First();
```
*Using a cast:*
``` csharp
var level = context.Levels
    .Include(l => l.Items).ThenInclude(c => ((Container)c).Items)
    .ToList().First();
```
Note that it would have been ideal if the compiler supported null conditional in LINQ expressions, so that the query could be written like this and it would be clear that the expression can be executed without throwing NullReferenceException or InvalidCastException:
``` csharp
 var level = context.Levels
    .Include(l => l.Items).ThenInclude(c => (c as Container)?.Items)
    .ToList().First();
```

However given the limitations in the compiler we believe the options above are good enough and can be interpreted as ways to address members that only exist on derived types without considering the potential for them to throw.

Possible advantages of this approach:

1. Eventually we can remove other restrictions from expressions used in call to `Include`/`ThenInclude`. Supporting cast/`as` is just a step in that direction.

2. This is what suggested by customers in https://data.uservoice.com/forums/72025-entity-framework-core-feature-suggestions/suggestions/1249289-eager-loading-for-properties-of-derived-classes, so we know it matches the expectations from some customers at least.

**Comment by divega** [comment]
Filed https://github.com/dotnet/csharplang/issues/1189.

**Comment by maumar** [comment]
fixed in a588722e10f34a12b9cea3d655fd26aafd91051b
